compiler/rustc_builtin_macros/src/assert/context.rs

```diff
@@ -0,0 +1,315 @@
+                        LitKind::Str(
+                            Symbol::intern(&format!("({})", cap.expr_str)),
+                            StrStyle::Cooked,
+                        ),
+                    ),
+                    self.cx.expr_ident(self.span, cap.var),
+                ],
+                self.span,
+            );
+
+            rustc_parse::nt_to_tokenstream(
+                &Nonterminal::NtExpr(new_call),
+                &self.cx.sess.parse_sess,
+                CanSynthesizeMissingTokens::Yes,
+            )
+            .into_trees()
+            .chain([TokenTree::token(token::Comma, self.span)])
+        });
+
+        self.cx.expr(
+            self.span,
+            ExprKind::MacCall(MacCall {
+                path: panic_path,
+                args: P(MacArgs::Delimited(
+                    DelimSpan::from_single(self.span),
+                    MacDelimiter::Parenthesis,
+                    initial_tts.into_iter().chain(captures_tts).collect::<TokenStream>(),
+                )),
+                prior_type_ascription: None,
+            }),
+        )
+    }
+
+    /// Tries to copy `__local_bind` into `__captureN`.
+    //
+    // {
+    //    let __local_bind = expr;
+    //    (&Wrapper(&__local_bind)).try_capture(&mut __captureN);
+    //    __local_bind
+    // }
+    fn build_try_capture(&self, capture: Ident, expr: &mut P<Expr>) {
+        let local_bind = Ident::new(sym::__local_bind, self.span);
+
+        let declaration = self.cx.stmt_let(vec![], self.span, false, local_bind, expr.clone());
+        let wrapper = self.cx.expr_call(
+            self.span,
+            self.cx.expr_path(self.cx.path(self.span, vec![Ident::with_dummy_span(sym::Wrapper)])),
+            vec![self.cx.expr_addr_of(self.span, self.cx.expr_path(Path::from_ident(local_bind)))],
+        );
+        let try_capture_call = self
+            .cx
+            .stmt_expr(self.cx.expr_method_call(
+                PathSegment {
+                    args: None,
+                    id: DUMMY_NODE_ID,
+                    ident: Ident::new(sym::try_capture, self.span),
+                },
+                vec![
+                    self.cx.expr_paren(self.span, self.cx.expr_addr_of(self.span, wrapper)),
+                    self.cx.expr_addr_of_mut(
+                        self.span,
+                        self.cx.expr_path(Path::from_ident(capture)),
+                    ),
+                ],
+                self.span,
+            ))
+            .add_trailing_semicolon();
+        let ret = self.cx.stmt_expr(self.cx.expr_path(Path::from_ident(local_bind)));
+        let stmts = vec![declaration, try_capture_call, ret];
+
+        *expr = self.cx.expr_block(self.cx.block(self.span, stmts));
+    }
+
+    /// Recursive function called until `cond_expr` is fully modified along side a string to
+    /// interpolate captured values.
+    fn manage_cond_expr(&mut self, cond_expr: &mut P<Expr>, fmt_str: &mut String) {
```

Comment:
This method is basically pretty printing with a custom additional logic. Perhaps it will be worth adding a custom `pprust::expr_to_string` function with `Path` callbacks? Something like `pprust::expr_to_string_cb(&Expr, impl FnMut(&mut Path))`?